jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java

```diff
@@ -0,0 +1,735 @@
+         * 
+         * @return the tour described by the list of vertex indices
+         */
+        public List<Integer> getTour()
+        {
+            return tour;
+        }
+
+        /**
+         * Returns true if all read vertices were distinct and non of them were
+         * {@link Object#equals(Object) equal}, else false.
+         * <p>
+         * The read {@link #getGraph() Graph} does not contain equal vertices because the vertices
+         * are stored in a {@link Set}. So in case the imported file contains duplicated locations
+         * all duplicates are discarded. This method tells if this was the case.
+         * </p>
+         * 
+         * @return true if only not equal vertices were imported from the file, else false
+         */
+        public boolean hasDistinctVertices()
+        {
+            return hasDistinctLocations;
+        }
+
+        /**
+         * Returns true if for each vertex all touching edges have different weights.
+         * <p>
+         * If this method returns true this means for the TSP that for each location each other
+         * location has a different distance, so there are no two other locations that have the same
+         * distance from that location.
+         * </p>
+         * 
+         * @return true if all touching edges of each vertex have different weight, else false
+         */
+        public boolean hasDistinctEdgesPerVertex()
+        {
+            if (hasDistinctEdges != null) {
+                return hasDistinctEdges;
+            }
+            hasDistinctEdges = Boolean.TRUE;
+
+            if (graph != null) {
+                for (V v : graph.vertexSet()) {
+                    Set<DefaultWeightedEdge> edgesOf = graph.edgesOf(v);
+
+                    Set<Double> weights = CollectionUtil.newHashSetWithExpectedSize(edgesOf.size());
+                    for (DefaultWeightedEdge edge : edgesOf) {
+                        weights.add(graph.getEdgeWeight(edge));
+                    }
+                    if (weights.size() != edgesOf.size()) {
+                        hasDistinctEdges = Boolean.FALSE;
+                        break;
+                    }
+                }
+            }
+            return hasDistinctEdges;
+        }
+    }
+
+    /**
+     * A getter to obtain element values of Vector data structures.
+     * 
+     * @author Hannes Wellmann
+     *
+     * @param <V> the type of handled vectors
+     */
+    public interface ElementGetter<V>
+    {
+        /**
+         * Returns the value of the element at the specified {@code index} in the given {@link V
+         * vector}.
+         * 
+         * @param vector from which an element value is to return
+         * @param index of the element to return
+         * @return the value of the specified element in the given vector
+         */
+        double getElement(V vector, int index);
+    }
+
+    private final Function<double[], V> vectorFactory;
```

Comment:
Completely remove the `vectorFactory` and the `elementGetter` abstractions. 

Graphs in JGraphT are build using the corresponding `vertexSupplier` inside the graph instance.

After you create the vertices using method `addVertex()`, you need to internally maintain a map from the vertices to the coordinates (read from file) of the vertex. Afterwards you need to iterate over all pair of vertices, acquire from the map the coordinates pair and call the corresponding distance method.

Reply:
All right. I agree with your remarks and will apply them in the next days.

The intention for the factory was to be able to set the coordinates read from the file to the vertices. I assumed that this importer would likely be used with some kind of vector data structure. But it looks like I was too specific here and I understand your point.

Nevertheless I would like to provide a user the possibility to have the read coordinates in the created vertices. What do you think about a `BiConsumer` to set the read coordinates into a vertex that a user can optionally supply in the constructor? The vertex would be created by the regular vertex supplier and the read coordinates would be stored Map as you suggested. Afterwards (if the setter is present) for each vertex its coordinates could be set to it with the setter. The subsequent edge computations would fully operate on the internal coordinate data structure of the importer.

Reply:
Or alternatively I provide the unmodifiable Vertex-Coordinates-Map in the MetaData-object. I think is better because simpler.